src/libstore/gc.cc

```diff
@@ -825,65 +817,87 @@ void LocalStore::collectGarbage(const GCOptions & options, GCResults & results)
             }
         } catch (GCLimitReached & e) {
         }
-    }
+    };
 
-    if (options.action == GCOptions::gcReturnLive) {
-        for (auto & i : alive)
-            results.paths.insert(printStorePath(i));
-        return;
-    }
+    /* Either delete all garbage paths, or just the specified
+       paths (for gcDeleteSpecific). */
+    std::visit(overloaded{
+        [&](GCDelete del){
+            if (del.pathsToDelete.has_value()) {
+                for (auto & i : del.pathsToDelete->paths) {
+                    deleteReferrersClosure(i);
+                    if (!del.pathsToDelete->skipAlive && !dead.count(i))
+                        throw Error(
+                            "Cannot delete path '%1%' since it is still alive. "
+                            "To find out why, use: "
+                            "nix-store --query --roots",
+                            printStorePath(i));
+                }
+            } else if (options.maxFreed > 0) {
```

Comment:
This means that `maxFreed` will be ignored unless we gc the whole store, right? Worth opening a follow-up issue about it then because that's probably not what we want if we use `--recursive`